### PR TITLE
Import BusyScreen Component

### PR DESCRIPTION
**Pull request by wackyleo459**
# Changes
 - BusyScreen.svelte
 - busy.store
 - busyState
 - page demonstration

# Screenshots
Screen Shot 2022-10-24 at 8 55 29 AM

### CODE REVIEW

**Comment by wackyleo459** [comment]
@peterpeterparker This branch still need some tweaks. There were some parts I wasn't sure about:
 - page demonstration `busyStore` section
 - Inside busyState, I know probably `BusyStateInitiatorType` should be kept in nns-dapp but not here?
 
 When you have time, can you leave some comments on this? Thank you!
 I also understand there would be relevant changes to nns-dapp for this, which I'll start working on.

File: src/lib/types/busyState.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  | "spawn-neuron"
+  | "stake-maturity"
+  | "claim_seed_neurons"
+  | "project-participate"
+  | "add-sns-hotkey-neuron"
+  | "remove-sns-hotkey-neuron"
+  | "disburse-neuron"
+  | "disburse-sns-neuron";
+
+export interface BusyState {
```
Comment (peterpeterparker) [inline src/lib/types/busyState.ts:30]: indeed `BusyStateInitiatorType` should remains in NNS-dapp

the state can maybe become generic?

```
export interface BusyState<T> {
   initiator: T;
   text?: string;
}
```

Comment (peterpeterparker) [reply src/lib/types/busyState.ts:30]: then I guess the store will have to be extended to support the generic type as well but maybe that works?

**Comment by peterpeterparker** [comment]
- nitpick: can you rename `busyState.ts` to `busy.state.ts` or `busy.ts` - i.e. lower case

- the store and component should be exported in `index.ts` and `components.ts`

File: src/routes/components/busy-screen/+page.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        initiator: initiator,
+        text: 'staking a neuron...',
+    });
+
+    const stopBusy = () => busyStore.stopBusy(initiator);
+
+    const click = (e) => {
+        e.preventDefault();
+        startBusy();
+        setTimeout(() => stopBusy(), 3000);
```
Comment (peterpeterparker) [inline src/routes/components/busy-screen/+page.md:18]: cool idea 👍

File: src/routes/components/busy-screen/+page.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+## Showcase
+
+<div>
+<button class="primary" id="start" on:click={click}>click to start Busy Screen</button>
+<BusyScreen />
+</div>
+
+<style>
+    .hello {
```
Comment (peterpeterparker) [inline src/routes/components/busy-screen/+page.md:53]: not sure, is that use anywhere?

File: src/routes/components/busy-screen/+page.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    }
+    
+</script>
+
+# BusyScreen
+
+A BusyScreen is a screen that displays a spinning loader and corresponding message (if available), while an application is loading content or performing another process. The component has an absolute centered positioning with the a faded overlay of the screen.
+
+```html
+import {busyStore} from "$lib/stores/busy.store"; import BusyScreen from
```
Comment (peterpeterparker) [inline src/routes/components/busy-screen/+page.md:28]: maybe you can follow the toast (https://ipi2f-uqaaa-aaaad-aabza-cai.ic0.app/components/toasts.html) pattern?

in html section the component and then another section usage that display how to use the store?

File: src/lib/stores/busy.store.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
+import { derived, writable, type Readable } from "svelte/store";
+import type { BusyState } from "$lib/types/busy";
+
+/**
+ * Store that reflects the app busy state.
+ * Is used to show the busy-screen that locks the UI.
+ */
+const initBusyStore = () => {
+  const { subscribe, update } = writable<Array<BusyState<any>>>([]);
```
Comment (peterpeterparker) [inline src/lib/stores/busy.store.ts:9]: we should not use `any`. we can replace it with the usage of generics?

```
const initBusyStore = <T>() => {
  const { subscribe, update } = writable<Array<BusyState<T>>>([]);
    ...
```

Then 

export const busyStore = initBusyStore();

File: src/lib/stores/busy.store.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+  return {
+    subscribe,
+    /**
+     * Show the busy-screen if not visible
+     */
+    startBusy({ initiator: newInitiator, text }: BusyState<any>) {
```
Comment (peterpeterparker) [inline src/lib/stores/busy.store.ts:16]: dito

```
startBusy({ initiator: newInitiator, text }: BusyState<T>) {
```

File: src/lib/stores/busy.store.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+      update((state) => [
+        ...state.filter(({ initiator }) => newInitiator !== initiator),
+        { initiator: newInitiator, text },
+      ]);
+    },
+
+    /**
+     * Hide the busy-screen if no other initiators are done
+     */
+    stopBusy(initiatorToRemove: BusyState<any>["initiator"]) {
```
Comment (peterpeterparker) [inline src/lib/stores/busy.store.ts:26]:
```
stopBusy(initiatorToRemove: T) {
```

File: src/lib/stores/busy.store.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+      update((state) =>
+        state.filter(({ initiator }) => initiator !== initiatorToRemove)
+      );
+    },
+  };
+};
+
+export const busyStore = initBusyStore();
```
Comment (peterpeterparker) [inline src/lib/stores/busy.store.ts:34]: ah I see now here we will have an issue with the generic...

you know what, maybe we can just replace it with a `string`

```
export interface BusyState {
  initiator: string;
  text?: string;
}
```

what do you think could that work? in NNS-dapp with use a type that contains `string` so maybe that will do?

**Comment by peterpeterparker** [comment]
@wackyleo459 we discussed with @mstrasinskis and we don't have better idea than using `string | number` instead of generic so, let's go for that if that works for you

File: src/lib/stores/busy.store.ts (hunk 8 above)
Comment (wackyleo459) [reply src/lib/stores/busy.store.ts:34]: I can make that change, seems most sensible

**Review by peterpeterparker — APPROVED** [review]
LGTM, thanks!

**Pull request merged**